sdk/azidentity/aad_identity_client.go

```diff
@@ -127,7 +164,32 @@ func (c *aadIdentityClient) createAccessToken(res *azcore.Response) (*azcore.Acc
 	// NOTE: look at go-autorest
 	accessToken.Token = value.Token
 	accessToken.ExpiresOn = time.Now().Add(time.Second * time.Duration(t)).UTC()
-	return accessToken, nil
+	return &tokenResponse{token: accessToken, refreshToken: value.RefreshToken}, nil
+}
+
+func (c *aadIdentityClient) createRefreshTokenRequest(tenantID, clientID, clientSecret, refreshToken string, scopes []string) (*azcore.Request, error) {
+	urlStr := c.options.AuthorityHost.String() + tenantID + tokenEndpoint
+	urlFormat, err := url.Parse(urlStr)
+	if err != nil {
+		return nil, err
+	}
+
+	data := url.Values{}
+	data.Set(qpGrantType, "refresh_token")
+	data.Set(qpClientID, clientID)
+	data.Set(qpClientSecret, clientSecret)
```

Comment:
@schaabs Is client secret required in all cases?  I see [this](https://docs.microsoft.com/en-us/azure/active-directory/develop/v1-protocols-oauth-code#refreshing-the-access-tokens) and it says only for web apps.

Reply:
I read that as well. This code  is going through the general refresh token flow. I believe it should be alright to send an empty value for client ID. I tested this with device code which does not send a client id and had no problems. However, I could simply add an if statement to check if clientID is empty. I'll wait to see what Scott says about this.

Reply:
Will add an if statement